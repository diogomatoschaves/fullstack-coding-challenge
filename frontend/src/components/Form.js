/**
 * Created by diogomatoschaves on 24/11/2018.
 */

import React, { Component } from 'react'
import { FormGroup, ControlLabel, FormControl, Button, ListGroupItem} from 'react-bootstrap'


class Form extends Component {

  static defaultProps = {
    sourceLang: 'en',
    targetLang: 'es'
  }

  state = {
    message: '',
    text: ''
  }

  getResult = ({ jobId, id, text }) => {

    const { addUid, addTranslation, updateTranslation, checkStatus, sourceLang, targetLang } = this.props
    let i = 0

    addTranslation({ id, jobId, originalText: text, sourceLang, targetLang, status: 'requesting' })

    this.setState({
      showMessage: true,
      message: 'Your request for translation was sent!',
      bsStyle: 'success'
    })
    setTimeout(() => {
      this.setState({ showMessage: false })
    }, 3000)

    const url = `${process.env.NODE_ENV === 'development' ? process.env.REACT_APP_API_URL_DEVELOPMENT 
      : process.env.REACT_APP_API_URL_PRODUCTION}/check_confirmation?jobId=${jobId}&id=${id}`;

    const headers = new Headers()

    const timeoutCallback = () => {
      i++
      fetch(url, {
        method: 'get',
        headers,
      })
        .then(res => res.json())
        .then(response => {
          if (response.uid) {
            addUid(id, response.uid)
            setTimeout(() => {
              checkStatus([{ id, uid: response.uid }])
            }, 3000)
          } else {
            if (response.status && response.status === 'failed') {
              updateTranslation(id, {
                status: 'failed',
              })
            } else {
              const { translations } = this.props
              if (Object.keys(translations).includes(id)) setTimeout(timeoutCallback, (7 - i) >= 2 ? (7 - i)*1000 : 2000)
            }
          }
        })
    }
    setTimeout(timeoutCallback, 7000)
  }

  buttonSubmit = () => {

    const { text } = this.state
    const { sourceLang, targetLang } = this.props

    this.setState({ text: '' })

    if (!text) return

    const headers = new Headers()
    
    const url = `${process.env.NODE_ENV === 'development' ? process.env.REACT_APP_API_URL_DEVELOPMENT 
      : process.env.REACT_APP_API_URL_PRODUCTION}/new_translation`;

    const body =  JSON.stringify({ text, sourceLang, targetLang })
    
    fetch(url, {
      method: 'post',
      headers,
      body
    })
      .then(res => res.json())
      .then(response => this.getResult({ jobId: response.job_id, id: response.translation_job, text }))
      .catch(error => {
        this.setState({
          showMessage: true,
          message: 'There was an error processing your request. Make sure you are connected to the internet.',
          bsStyle: 'danger'
        })
        setTimeout(() => {
          this.setState({ showMessage: false })
        }, 3000)
      })
  }

  handleChange = (e) => {
    this.setState({ text: e.target.value });
  }
  
  render() {

    const { text, message, showMessage, bsStyle } = this.state
    
    return (
      <div className="flex-row" style={{width: '80%', maxWidth: '800px', justifyContent: 'space-around'}}>
        <FormGroup style={{width: '77%'}}>
          <ControlLabel>Text to be translated</ControlLabel>
          <FormControl
            value={text}
            onChange={this.handleChange}
            componentClass="textarea"
            placeholder="Insert your text here"
            style={{height: '150px'}}
          />
        </FormGroup>
        <div className="flex-row" style={{width: '20%'}}>
          <Button disabled={!text} bsStyle="primary" bsSize="large" onClick={this.buttonSubmit}>
            Request
          </Button>
        </div>
        <ListGroupItem
          className='message-to-user flex-column'
          bsSize="large"
          bsStyle={bsStyle}
          style={{top: showMessage ? '60px' : '-200px'}}
        >
          {message}
        </ListGroupItem>
      </div>
    )
  }
}

export default Form
