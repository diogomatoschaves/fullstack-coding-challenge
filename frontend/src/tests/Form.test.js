/**
 * Created by diogomatoschaves on 26/11/2018.
 */

import { shallow } from 'enzyme'
import React from 'react'
import Form from '../Form'


it('expect to render Form component', () => {
  
  const mockSourceLangs = [{
      name: 'English',
      key: 'en'
    }]
  
  const mockTargetLangs = [{
      name: 'Spanish',
      key: 'es'
    }]

  expect(shallow(<Form sourceLangs={mockSourceLangs} targetLangs={mockTargetLangs}/>)).toMatchSnapshot()
  
})
