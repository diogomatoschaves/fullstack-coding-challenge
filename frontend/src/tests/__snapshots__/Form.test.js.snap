// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`expect to render Form component 1`] = `
<div
  className="flex-column"
  style={
    Object {
      "justifyContent": "space-around",
      "maxWidth": "850px",
      "width": "80%",
    }
  }
>
  <div
    className="flex-row"
    style={
      Object {
        "alignSelf": "flex-start",
        "justifyContent": "flex-start",
      }
    }
  >
    <div>
      <strong>
        From:
      </strong>
      <DropdownButton
        id="dropdown-basic"
        style={
          Object {
            "marginLeft": "10px",
          }
        }
        title="English"
      >
        <MenuItem
          bsClass="dropdown"
          disabled={false}
          divider={false}
          eventKey="1"
          header={false}
          key="en"
          onClick={[Function]}
        >
          English
        </MenuItem>
      </DropdownButton>
    </div>
    <div
      style={
        Object {
          "paddingLeft": "20px",
        }
      }
    >
      <strong>
        To:
      </strong>
      <DropdownButton
        id="dropdown-basic"
        style={
          Object {
            "marginLeft": "10px",
          }
        }
        title="Spanish"
      >
        <MenuItem
          bsClass="dropdown"
          disabled={false}
          divider={false}
          eventKey="1"
          header={false}
          key="es"
          onClick={[Function]}
        >
          Spanish
        </MenuItem>
      </DropdownButton>
    </div>
  </div>
  <div
    className="flex-row"
    style={
      Object {
        "justifyContent": "space-between",
        "marginTop": "15px",
        "width": "100%",
      }
    }
  >
    <FormGroup
      bsClass="form-group"
      style={
        Object {
          "width": "77%",
        }
      }
    >
      <FormControl
        bsClass="form-control"
        componentClass="textarea"
        onChange={[Function]}
        placeholder="Insert your text here"
        style={
          Object {
            "height": "150px",
          }
        }
        value=""
      />
    </FormGroup>
    <div
      className="flex-row"
      style={
        Object {
          "width": "20%",
        }
      }
    >
      <Button
        active={false}
        block={false}
        bsClass="btn"
        bsSize="large"
        bsStyle="primary"
        disabled={true}
        id="send-request"
        onClick={[Function]}
      >
        Send Request
      </Button>
    </div>
  </div>
  <ListGroupItem
    bsClass="list-group-item"
    bsSize="large"
    className="message-to-user flex-column"
    listItem={false}
    style={
      Object {
        "top": "-200px",
      }
    }
  />
</div>
`;
